penguin/web/sse_events.py

```
@@ -305,55 +342,64 @@ def event_handler(event_type: str, data: Any):
+            # Ledger failures must not interrupt live TUI streaming. The client can
+            # still receive the live frame; replay will surface a gap if needed.
+            logger.debug(
+                "Failed to record OpenCode event in runtime ledger",
+                exc_info=True,
+            )
+
+    subscribe("opencode_event", ledger_handler)
+    setattr(core, _LEDGER_RECORDER_ATTR, ledger_handler)
 
 
 def _replay_events_after(
```

Ruff (0.15.20)

[warning] Missing return type annotation for private function `_replay_events_after`

(ANN202)